hypershift-operator/controllers/hostedcluster/internal/platform/core.go

```diff
@@ -0,0 +1,60 @@
+package platform
+
+import (
+	"context"
+	"fmt"
+
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"github.com/openshift/hypershift/support/upsert"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	hostedClusterAnnotation = "hypershift.openshift.io/cluster"
+)
+
+type Platform interface {
+	// ReconcileCAPIInfraCR is responsible for creating or updating a CAPIInfraCR.
+	ReconcileCAPIInfraCR(hcluster *hyperv1.HostedCluster, name, controlPlaneNamespace string, apiEndpoint hyperv1.APIEndpoint,
+		c client.Client, ctx context.Context) (client.Object, error)
```

Comment:
The contract here isn't entirely clear to me. The user is given some input state and a client which is apparently (based on the current implementations) allowed to do arbitrary mutations and then returns some state (representing what?)

What is the user allowed or expected to do, especially with the client? What could the reconciler be doing on behalf of the user to make it more clear what the contract is? At what point in the overall lifecycle is this invoked?

An example for consideration:

```go
// ReconcileCAPIInfraCR is called during HostedCluster reconciliation prior to HostedCluster status calculations.
// Implementations should use the given input and read-only client to return the new desired state of the
// provider's CAPI CR resource, which will then be immediately persisted.
```